File: notify/R/notifyclient.R
# Copyright Core DF — Apache License 2.0

notify_missing_env <- function(vars) list(status_code = 601L, error = paste("Environment variables", vars, "should be defined"))
notify_transport_error <- function(message = "inaccessible") list(status_code = 0L, error = message)

`%||%` <- function(a, b) if (is.null(a) || (is.character(a) && !nzchar(a))) b else a

.post_json <- function(url, payload) {
  if (!requireNamespace("httr", quietly = TRUE)) stop("install.packages('httr')")
  if (!requireNamespace("jsonlite", quietly = TRUE)) stop("install.packages('jsonlite')")
  resp <- tryCatch(
    httr::POST(
      url,
      httr::add_headers(`Content-Type` = "application/json"),
      body = jsonlite::toJSON(payload, auto_unbox = TRUE),
      encode = "raw"
    ),
    error = function(e) NULL
  )
  if (is.null(resp)) return(notify_transport_error())
  code <- as.integer(httr::status_code(resp))
  text <- httr::content(resp, as = "text", encoding = "UTF-8")
  if (code >= 400L) return(list(status_code = code, error = text))
  if (!nzchar(text)) return(list(status_code = 200L))
  body <- tryCatch(jsonlite::fromJSON(text, simplifyVector = FALSE), error = function(e) text)
  list(status_code = 200L, body = body)
}

Slack <- function(text, webhook_url = NULL) {
  url <- webhook_url %||% Sys.getenv("SLACK_WEBHOOK_URL")
  if (!nzchar(url)) return(notify_missing_env("SLACK_WEBHOOK_URL"))
  .post_json(url, list(text = text))
}

Teams <- function(text, webhook_url = NULL) {
  url <- webhook_url %||% Sys.getenv("TEAMS_WEBHOOK_URL")
  if (!nzchar(url)) return(notify_missing_env("TEAMS_WEBHOOK_URL"))
  .post_json(url, list(
    "@type" = "MessageCard",
    "@context" = "http://schema.org/extensions",
    text = text
  ))
}

PagerDuty <- function(summary, routing_key = NULL, severity = "error") {
  key <- routing_key %||% Sys.getenv("PAGERDUTY_ROUTING_KEY")
  if (!nzchar(key)) return(notify_missing_env("PAGERDUTY_ROUTING_KEY"))
  .post_json("https://events.pagerduty.com/v2/enqueue", list(
    routing_key = key,
    event_action = "trigger",
    payload = list(summary = summary, severity = severity, source = "coreauto-step")
  ))
}

Email <- function(subject, body, to_addrs, from_addr = NULL) {
  host <- Sys.getenv("SMTP_HOST")
  port <- as.integer(Sys.getenv("SMTP_PORT", "587"))
  user <- Sys.getenv("SMTP_USER")
  password <- Sys.getenv("SMTP_PASSWORD")
  sender <- from_addr %||% Sys.getenv("SMTP_FROM") %||% user
  if (!nzchar(host) || !nzchar(sender)) return(notify_missing_env("SMTP_HOST and SMTP_FROM (or from_addr)"))
  tryCatch({
    if (!requireNamespace("emayili", quietly = TRUE)) stop("install.packages('emayili')")
    smtp <- emayili::server(host = host, port = port, username = user, password = password, tls = nzchar(user))
    msg <- emayili::envelope(from = sender, to = to_addrs, subject = subject) %>% emayili::text(body)
    emayili::smtp_send(msg, smtp)
    list(status_code = 200L)
  }, error = function(e) notify_transport_error(e$message))
}

File: notify/DotNet/NotifyClient.cs
// Copyright Core DF
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System.Net.Http.Json;
using System.Net.Mail;
using System.Text.Json;

namespace CoreAuto.Notify;

public static class NotifyClient
{
    private static readonly HttpClient Http = new() { Timeout = TimeSpan.FromSeconds(30) };

    public static async Task<Result> SlackAsync(string text, string? webhookUrl = null)
    {
        var url = webhookUrl ?? Env("SLACK_WEBHOOK_URL");
        if (string.IsNullOrEmpty(url)) return Result.MissingEnv("SLACK_WEBHOOK_URL");
        return await PostJsonAsync(url, new { text });
    }

    public static async Task<Result> TeamsAsync(string text, string? webhookUrl = null)
    {
        var url = webhookUrl ?? Env("TEAMS_WEBHOOK_URL");
        if (string.IsNullOrEmpty(url)) return Result.MissingEnv("TEAMS_WEBHOOK_URL");
        return await PostJsonAsync(url, new { type = "MessageCard", context = "http://schema.org/extensions", text });
    }

    public static Result Email(string subject, string body, string toAddrs, string? fromAddr = null)
    {
        var host = Env("SMTP_HOST"); var port = int.Parse(Env("SMTP_PORT", "587"));
        var user = Env("SMTP_USER"); var password = Env("SMTP_PASSWORD");
        var sender = fromAddr ?? Env("SMTP_FROM", user);
        if (string.IsNullOrEmpty(host) || string.IsNullOrEmpty(sender)) return Result.MissingEnv("SMTP_HOST and SMTP_FROM (or from_addr)");
        try {
            using var client = new SmtpClient(host, port) { EnableSsl = true, Credentials = string.IsNullOrEmpty(user) ? null : new System.Net.NetworkCredential(user, password) };
            using var msg = new MailMessage(sender, toAddrs.Split(',')[0].Trim(), subject, body);
            foreach (var a in toAddrs.Split(',').Skip(1)) msg.To.Add(a.Trim());
            client.Send(msg); return Result.Ok();
        } catch (Exception ex) { return Result.TransportError(ex.Message); }
    }

    public static async Task<Result> PagerDutyAsync(string summary, string? routingKey = null, string severity = "error")
    {
        var key = routingKey ?? Env("PAGERDUTY_ROUTING_KEY");
        if (string.IsNullOrEmpty(key)) return Result.MissingEnv("PAGERDUTY_ROUTING_KEY");
        var payload = new { routing_key = key, event_action = "trigger", payload = new { summary, severity, source = "coreauto-step" } };
        return await PostJsonAsync("https://events.pagerduty.com/v2/enqueue", payload);
    }

    private static async Task<Result> PostJsonAsync(string url, object body)
    {
        try {
            using var resp = await Http.PostAsJsonAsync(url, body);
            var text = await resp.Content.ReadAsStringAsync();
            if ((int)resp.StatusCode >= 400) return Result.Error((int)resp.StatusCode, text);
            var extra = new Dictionary<string, object?>();
            if (!string.IsNullOrWhiteSpace(text)) extra["body"] = JsonSerializer.Deserialize<JsonElement>(text);
            return Result.Ok(extra.Count > 0 ? extra : null);
        } catch (Exception ex) { return Result.TransportError(ex.Message); }
    }

    private static string Env(string k, string d = "") => Environment.GetEnvironmentVariable(k) ?? d;
}
